File: src-1.x/BYTES.NET/WPF/MVVM/ViewModelRelayCommand.vb
Namespace WPF.MVVM

    ''' <summary>
    ''' base class for a flexible view model relay command
    ''' </summary>
    ''' <remarks>based on the article found at 'http://www.cocktailsandcode.de/2012/04/mvvm-tutorial-part-3-viewmodelbase-und-relaycommand/'</remarks>
    Public Class ViewModelRelayCommand

#Region "method(s) implementing 'ICommand' interface"

        Implements ICommand

        Public Event CanExecuteChanged(sender As Object, e As EventArgs) Implements ICommand.CanExecuteChanged

        ''' <summary>
        ''' method for retrieving a boolean indicating wheather a the command can be executed or not
        ''' </summary>
        ''' <param name="parameter"></param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Function CanExecute(parameter As Object) As Boolean Implements ICommand.CanExecute

            Return _canExecute

        End Function

        ''' <summary>
        ''' method for executing the command
        ''' </summary>
        ''' <param name="parameter"></param>
        ''' <remarks></remarks>
        Public Sub Execute(parameter As Object) Implements ICommand.Execute

            If Not IsNothing(_action) Then
                _action(parameter)
            Else

                If Not IsNothing(_parameterLessAction) Then
                    _parameterLessAction()
                End If

            End If

        End Sub

#End Region

#Region "private variable(s)"

        Private ReadOnly _action As Action(Of Object) = Nothing
        Private ReadOnly _parameterLessAction As Action = Nothing

        Private _canExecute As Boolean = Nothing

#End Region

#Region "public poperties"

        ''' <summary>
        ''' property determining wheather the command can be executed or not
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property IsEnabled As Boolean
            Get

                Return _canExecute

            End Get
            Set(value As Boolean)

                _canExecute = value
                Me.OnCanExecuteChanged()

            End Set
        End Property

#End Region

#Region "public new instance method(s)"

        ''' <summary>
        ''' overloaded new instance method
        ''' </summary>
        ''' <param name="action"></param>
        ''' <param name="canExecute"></param>
        ''' <remarks></remarks>
        Public Sub New(action As Action(Of Object), Optional ByVal canExecute As Boolean = True)

            'set the variable(s)
            _action = action
            _canExecute = canExecute

        End Sub

        ''' <summary>
        ''' overvloaded new instance method, accepting a parameter-less action
        ''' </summary>
        ''' <param name="action"></param>
        ''' <param name="canExecute"></param>
        ''' <remarks></remarks>
        Public Sub New(action As Action, Optional ByVal canExecute As Boolean = True)

            'set the variable(s)
            _parameterLessAction = action
            _canExecute = canExecute

        End Sub

#End Region

#Region "public method(s)"

        ''' <summary>
        ''' method for updating the 'CanExecute' status
        ''' </summary>
        ''' <remarks></remarks>
        Public Sub OnCanExecuteChanged()

            RaiseEvent CanExecuteChanged(Me, EventArgs.Empty)

        End Sub

#End Region

    End Class

End Namespace

File: src-2023/BYTES.NET.WPF/MVVM/ViewModelRelayCommand.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Input;

namespace BYTES.NET.WPF.MVVM
{
    /// <summary>
    /// the basic MVVM view model relay command type
    /// </summary>
    /// <remarks>based on the article found at 'http://www.cocktailsandcode.de/2012/04/mvvm-tutorial-part-3-viewmodelbase-und-relaycommand/'</remarks>
    public class ViewModelRelayCommand : ICommand
    {
        #region private variable(s)

        private bool? _canExecuteBool = null;
        private Predicate<object> _canExecuteDelegate = null;

        private Action _toBeExecutedWithoutParameter = null;
        private Action<object> _toBeExecutedWithParameter = null;

        private bool _executeAsync = false;

        #endregion

        #region public properties

        public bool RunAsync
        {
            get => _executeAsync;
            set => _executeAsync = value;
        }

        public bool IsEnabled
        {
            get => CanExecute(null);
            set
            {
                _canExecuteBool = value;
                OnCanExecuteChanged();
            }
        }

        #endregion

        #region public new instance method(s)

        /// <summary>
        /// default new instance method
        /// </summary>
        /// <param name="toBeExecuted"></param>
        /// <param name="canExecute"></param>
        public ViewModelRelayCommand(Action toBeExecuted, bool canExecute = true)
        {
            _toBeExecutedWithoutParameter = toBeExecuted;
            _canExecuteBool = canExecute;
        }

        /// <summary>
        /// overloaded new instance method, supporting an 'object' type parameter for the delegated 'Action'
        /// </summary>
        /// <param name="toBeExecuted"></param>
        /// <param name="canExecute"></param>
        public ViewModelRelayCommand(Action<object> toBeExecuted, bool canExecute = true)
        {
            _toBeExecutedWithParameter = toBeExecuted;
            _canExecuteBool = canExecute;
        }

        /// <summary>
        /// overloaded new instance method, supporting a predicate parameter for calculating 'CanExecute'
        /// </summary>
        /// <param name="toBeExecuted"></param>
        /// <param name="canExecute"></param>
        public ViewModelRelayCommand(Action toBeExecuted, Predicate<object> canExecute)
        {
            _toBeExecutedWithoutParameter = toBeExecuted;
            _canExecuteDelegate = canExecute;
        }

        /// <summary>
        /// overloading new instance method, supporting an 'object' type parameter for the delegated 'Action' and a predicate parameter for calculating 'CanExecute'
        /// </summary>
        /// <param name="toBeExecuted"></param>
        /// <param name="canExecute"></param>
        public ViewModelRelayCommand(Action<object> toBeExecuted, Predicate<object> canExecute)
        {
            _toBeExecutedWithParameter = toBeExecuted;
            _canExecuteDelegate = canExecute;
        }

        #endregion

        #region public event(s) and method(s) implementing 'ICommand'

        public event EventHandler CanExecuteChanged;

        public bool CanExecute(object parameter)
        {
            if (_canExecuteBool.HasValue)
            {

                if (_canExecuteBool != null)
                {
                    return (bool)_canExecuteBool;
                }

            }
            else
            {

                if (_canExecuteDelegate != null)
                {
                    return _canExecuteDelegate(parameter);
                }

            }

            //return the default value
            return true;
        }

        public void Execute(object parameter)
        {
            if (this.RunAsync)
            {
                ExecuteAsync();
            }
            else
            {
                if (_toBeExecutedWithoutParameter != null)
                {
                    _toBeExecutedWithoutParameter();
                }
                else if (_toBeExecutedWithParameter != null)
                {
                    _toBeExecutedWithParameter(parameter);
                }
            }
        }

        #endregion

        #region private method(s) supporting 'ICommand'

        /// <summary>
        /// calling the 'CanExecuteChanged' event on changing the 'CanExecute' property
        /// </summary>
        private void OnCanExecuteChanged()
        {
            CanExecuteChanged(this, EventArgs.Empty);
        }

        /// <summary>
        /// running the delegated action async
        /// </summary>
        private async void ExecuteAsync()
        {

            if (_toBeExecutedWithoutParameter != null)
            {
                await Task.Run(_toBeExecutedWithoutParameter);
            }
            else if (_toBeExecutedWithParameter != null)
            {
                await Task.Run(() => _toBeExecutedWithParameter);
            }

        }

        #endregion
    }
}
